File: realm-common/src/main/java/io/github/solusmods/eternalcore/realm/impl/RealmTagProvider.java
package io.github.solusmods.eternalcore.realm.impl;

import io.github.solusmods.eternalcore.realm.api.Realm;
import net.minecraft.core.HolderLookup;
import net.minecraft.data.PackOutput;
import net.minecraft.data.tags.IntrinsicHolderTagsProvider;

import java.util.concurrent.CompletableFuture;

public abstract class RealmTagProvider extends IntrinsicHolderTagsProvider<Realm> {
    public RealmTagProvider(PackOutput packOutput, CompletableFuture<HolderLookup.Provider> completableFuture) {
        super(packOutput, RealmRegistry.KEY, completableFuture, realm -> RealmRegistry.REALMS.getKey(realm).orElseThrow());
    }

    public RealmTagProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider, CompletableFuture<TagLookup<Realm>> parentProvider) {
        super(output, RealmRegistry.KEY, lookupProvider, parentProvider, realm -> RealmRegistry.REALMS.getKey(realm).orElseThrow());
    }

    @Override
    protected void addTags(HolderLookup.Provider provider) {

    }
}

File: realm-common/src/main/kotlin/io/github/solusmods/eternalcore/realm/impl/RealmTagProvider.kt
package io.github.solusmods.eternalcore.realm.impl

import io.github.solusmods.eternalcore.realm.api.Realm
import net.minecraft.core.HolderLookup
import net.minecraft.data.PackOutput
import net.minecraft.data.tags.IntrinsicHolderTagsProvider
import java.util.concurrent.CompletableFuture
import java.util.function.Function

abstract class RealmTagProvider : IntrinsicHolderTagsProvider<Realm?> {
    constructor(packOutput: PackOutput, completableFuture: CompletableFuture<HolderLookup.Provider?>?) : super(
        packOutput,
        RealmRegistry.KEY,
        completableFuture,
        Function { realm: Realm? -> RealmRegistry.REALMS.getKey(realm).orElseThrow() })

    constructor(
        output: PackOutput,
        lookupProvider: CompletableFuture<HolderLookup.Provider?>?,
        parentProvider: CompletableFuture<TagLookup<Realm?>?>?
    ) : super(
        output,
        RealmRegistry.KEY,
        lookupProvider,
        parentProvider,
        Function { realm: Realm? -> RealmRegistry.REALMS.getKey(realm).orElseThrow() })
}
